src/grafana_agent.py

```diff
@@ -700,26 +795,109 @@ def _integrations_config(self) -> dict:
+            a dict with the receivers config.
+        """
+        if not self._tracing.is_ready():
+            return {}
+
+        receivers_set = self.requested_receivers
+
+        if not receivers_set:
+            logger.warning("No tempo receivers enabled: grafana-agent cannot ingest traces.")
+            return {}
+
+        if self.cert.enabled:
+            base_receiver_config: Dict[str, Union[str, Dict]] = {
+                "tls": {
+                    "ca_file": str(self._ca_path),
+                    "cert_file": str(self._cert_path),
+                    "key_file": str(self._key_path),
+                    "min_version": "",
+                }
+            }
+        else:
+            base_receiver_config = {}
+
+        def _receiver_config(protocol: str):
+            endpoint = "0.0.0.0:" + str(self._tracing_receivers_ports[protocol])  # type: ignore
+            receiver_config = base_receiver_config.copy()
+            receiver_config["endpoint"] = endpoint
+            return receiver_config
+
+        config = {}
+
+        if "zipkin" in receivers_set:
+            config["zipkin"] = _receiver_config("zipkin")
+        if "opencensus" in receivers_set:
+            config["opencensus"] = _receiver_config("opencensus")
+
+        otlp_config = {}
+        if "otlp_http" in receivers_set:
+            otlp_config["http"] = _receiver_config("otlp_http")
+        if "otlp_grpc" in receivers_set:
+            otlp_config["grpc"] = _receiver_config("otlp_grpc")
+        if otlp_config:
+            config["otlp"] = {"protocols": otlp_config}
+
+        jaeger_config = {}
+        if "jaeger_thrift_http" in receivers_set:
+            jaeger_config["thrift_http"] = _receiver_config("jaeger_thrift_http")
+        if "jaeger_grpc" in receivers_set:
+            jaeger_config["grpc"] = _receiver_config("jaeger_grpc")
+        if jaeger_config:
+            config["jaeger"] = {"protocols": jaeger_config}
+
+        return config
+
+    @property
+    def _tempo_config(self) -> Dict[str, Union[Any, List[Any]]]:
+        """The tracing section of the config.
+
+        Returns:
+            a dict with the tracing config.
+        """
+        endpoints = self._endpoints_with_tls()
+        receivers = self._tracing_receivers
+
+        if not receivers:
+            # pushing a config with an empty receivers section will cause gagent to error out
+            return {}
+
+        return {
+            "configs": [
+                {
+                    "name": "tempo",
+                    "remote_write": endpoints.tempo,
+                    "receivers": receivers,
+                }
+            ]
+        }
+
     @property
     def _loki_config(self) -> Dict[str, Union[Any, List[Any]]]:
         """Modifies the loki section of the config.
 
         Returns:
             a dict with Loki config
         """
-        _, loki_endpoints = self._enrich_endpoints()
+        endpoints = self._endpoints_with_tls()
```

Comment:
I don't like that we're calling this from three different places, every time to only use a different third of the information it computes.

proposals (sorted by bother):
- store it in an instance attribute in `__init__`, so we don't recompute on each access
- split it in three, and call only the specific one whenever we need it
- parametrize with the subset we're interested in, and only generate that one
- lazify computation on access

Reply:
I also don't see the point of it being bundled together anymore. I think there was an idea here but it doesn't seem to work out as nice as it seemed at the beginning so I've done the split into three. I also don't see the purpose of the dataclass that was created for it anymore so I also removed the dataclass. Let me know if you think it went too far.